[28기 이가윤] 리팩토링 중

Gayun00 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [ ] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 아이디, 비밀번호 입력창을 사용한 로그인 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 입력 데이터 유효성 검사
- 입력창에 변화가 일어날 때 마다 유효성 검사 함수 실행
- validState 값(boolean)에 따라 disabled, 버튼 색상 class 변경, 주의 메시지 조건부 렌더링


2. 메인 화면 댓글 삭제/추가 기능
- Feed 컴포넌트에서 addComment, deleteComment   함수 정의
- Comments > Comment 컴포넌트로 전달 후 state 끌어올려 댓글 변경사항 적용

3. 무한 스크롤, 스켈레톤 로딩 기능
- Feeds 컴포넌트에서 loadFeedData, loadMoreFeed, createSkeletonFeed() 함수 정의
- Feeds 컴포넌트 최하단에 feedEnd 요소 생성 / observe
- 교차할 때마다 5개의 새로운 피드 데이터 불러오기



<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
   - 무한 스크롤을 구현하는 과정에서 throttle과 raf를 사용해 intersectionObserver을 최적화 시켜줄 수 있다는 것을 알게되었습니다. 시간이 남는다면 한번 시도해보고 싶습니다.
   - scss를 다루는 방법에 대해 알게 되었습니다.
   - 불필요한 리렌더링을 막기 위해 React.memo, useCallback, useMemo를 사용할 수 있다는 것을 알게되었습니다. 이 부분도 react dev tool을 사용해 불필요한 리렌더링이 발생하는 부분을 찾아 개선해보고 싶습니다.

<br />

## :: 기타 질문 및 특이 사항
- state를 불필요하게 많이 만들어 관리하고 있는 것 같다는 생각이 들어서, 굳이 state을 만들어 관리할 필요가 없는 부분이 있는지 궁금합니다.
- 상단 story 영역에 필요한 데이터는 MainGayun에서 fetch 후 하위 props로 넘겨주었고,
피드 영역에 필요한 데이터는 Feeds 컴포넌트에서 fetch해 불러왔습니다. 피드와 관련된 주요 데이터는 피드 컴포넌트 안에서 관리하는게 더 낫지 않을까 하는 생각이었는데, 최상위 컴포넌트에서 일관되게 모든 데이터를 fetch 한 뒤 각 하위 컴포넌트로 넘겨주는 방식이 더 깔끔하고 관리하기 좋은 방법일까요?
- intersectionObserver 부분은 커스텀 훅으로 다시 만들어 보는 중입니다.

Gayun00 committed:
[Add] branch 생성

Gayun00 committed:
test

Gayun00 committed:
[Add] Login, Main-Feed.js 컴포넌트 추가

Gayun00 committed:
[Add] Login, Main-Feed 컴포넌트 추가
Merge branch 'master' into feature/gayun

Gayun00 committed:
commit test.2

Gayun00 committed:
commit test.5

Gayun00 committed:
[Add] 모든 컴포넌트 추가
- css 수정중

Gayun00 committed:
[Modify] css 수정 진행중
- Story 컴포넌트 수정
- 스토리 이미지 배경 추가
- 피드 아이콘 마진 추가

Gayun00 committed:
[Modify] SideBar:  css 수정 진행중

Gayun00 committed:
[Modify] 로고 폰트 추가

Gayun00 committed:
[Add] SideBar: account 영역 추가

Gayun00 committed:
[Add] LoginGayun: Footer 공통 컴포넌트 추가

Gayun00 committed:
[Modify] LoginGayun css 수정완료
- container 추가하여 로그인 화면 감싸는 테두리 추가
- 어긋난 요소 재배치
- 첫 화면 notice 메시지 보이지 않도록 state 초기값 false로 설정

Gayun00 committed:
[Modify] Story, Feed, Nav: css 수정
- story 위 공백 추가
- feed 인풋 버튼 색상 변경
- nav바 넓이 변경

Gayun00 committed:
[Add] Feed,Comments,Comment: 댓글 추가

Gayun00 committed:
[Add] Mission1 - 사용자 입력 데이터 저장 기능 구현

Gayun00 committed:
[Add] Mission 2 - 로그인 버튼 활성화 기능 구현

Gayun00 committed:
question

Gayun00 committed:
[Add] Mission 3 - 댓글 기능 구현 (추가)

Gayun00 committed:
[Add] 로그인 버튼 활성화/비활성화 기능 추가

Gayun00 committed:
[Add] LoginGayun: 로그인 버튼활성화 / 클릭시 페이지 이동
- useNavigate로 버튼 클릭시 메인화면으로 이동
- 로그인 유효성 검사 통과시 버튼 활성화 : disabled 수정

Gayun00 committed:
[Add] Feeds, Comments, Story, SideBarkey props 추가

Gayun00 committed:
[Add] Feeds: feed.json 비동기로 받아오기

Gayun00 committed:
[Modify] SideBar:  반응형 구현
- 브라우저 창 너비 축소시 SideBar 사라지도록 구현 (클래스 추가)

Gayun00 committed:
[Fix] Feeds:  무한 스크롤 기능 오류 수정
- 스크롤 끝 도달 시 새로 생성된 피드의 최하단으로 스크롤 고정되어 있는 오류 수정
- observer 부착된 feedEnd 요소 feeds-container 밖으로 빼 최하단으로 배치

Gayun00 committed:
[Add] Skeleton: Skeleton 컴포넌트 추가

Gayun00 committed:
[Add] Skeleton: loading 화면 추가
- img-content 요소 내부에 loading 요소 추가
- keyframes로 무한 회전하는 로딩 요소 구현

Gayun00 committed:
[Add] Feeds: feedEnd 요소 로딩 화면 추가
- 최초 렌더링 시 스켈레톤 화면
- 이후 추가 피드 렌더링 전 피드 최하단에 loading 화면 표시

Gayun00 committed:
[Fix] Feeds: 최초 렌더링 시 스켈레톤 피드 복수 생성
- createSkeletonFeed() 추가

Gayun00 committed:
[Fix] Feeds: 최초 렌더링 시 스켈레톤 피드, 스크롤 끝 도달 시 추가 피드 로드 오류 수정완료

Gayun00 committed:
[Modify] 파일 위치 수정, class > className 수정

Gayun00 committed:
클래스 수정

Gayun00 committed:
[Fix] Story, SideBar: data fetch 수정

Gayun00 committed:
[Add] LoginGayun:  로그인 백엔드 통신 추가

Gayun00 committed:
[Modify] Comment: 댓글 css tnwjd

minami-cs commented on src/components/Footer/Footer.js:
```diff
@@ -1,25 +1,49 @@
+            <a href="https://wecode.co.kr/">Wecode</a>
+          </li>
+          <li>
+            <a>소개</a>
```
음 제가 footer에 쓸데없이 a태그를 넣어놔섴ㅋㅋㅋㅋㅋㅋ이러게 되어버렸네요... 원래 하나씩 진짜 인스타 footer의 링크를 걸려고 했는데 여태 잊고 있었습미다

Gayun00 committed:
[Add] UseInfiniteScroll: 무한 스크롤 기능 custom hooks로 별도 생성 중

Gayun00 committed:
[Refactor] UseInfiniteScroll 수정 중
- useEffect 내부 observe & unobserve 로직 완성

ghost commented:
Sider has detected **1** warning on analyzing the commit 9d06981.

- ⚠️ The number of issues (56) exceeded the limit of comments. Please visit [Sider](https://sider.review/gh/repos/438165292/pulls/3) to review the issues.

If the errors persist even after [retrying](https://sider.review/gh/repos/438165292/pulls/3#tools), the following actions may resolve them:
- Update tool dependencies
- Configure failed tools
- [Configure `sider.yml`](https://sider.review/gh/repos/438165292/pulls/3#config)
- [Turn off unused tools](https://sider.review/gh/repos/438165292/settings)

If you still have problems, feel free to ask us via chat. 💬

---

You can [turn off such notifications](https://sider.review/gh/repos/438165292/settings#notifications) if unnecessary.

Gayun00 committed:
[Add] LoadMoreFeed 컴포넌트 추가

Gayun00 committed:
[Refactor] Feeds: 무한 스크롤 기능 리팩토링 중
- state 수정 : isLoading, page, feedList
- page 값을 기준으로 fetch하도록 useEffect()내부 setIsLoading, fetchFeedData() 추가
- fetchFeedData() 안에서 page값에 따른 데이터를 불러오도록 변경

Gayun00 committed:
[Refactor] LoadMoreFeed: intersectionObserver 영역 추가
- useEffect() 내부에서 마운트/언마운트시 observe/unobserve 설정
- 전달받은 setPage()로 스크롤 끝 도달 시 page + 1

Gayun00 committed:
[Fix] 최초 렌더링 시 피드가 한번 더 생성되는 오류 수정
- 최초 렌더링 시 intersectionObserver 작동
 - 불러올 피드가 없어 최하당 feedEnd 요소가 위로 끌어올려지면서, isIntersecting && intersectionRatio === 1 조건을 충족한 것이 원인
- 이를 해결해줄 skeleton feed가 제대로 렌더링 되지 않음
 - feedData를 fetch하기 전 loading이 true > false로 즉시 변경되는 것이 원인
- fetchFeedData를 비동기 처리해 피드로드가 끝난 후에 로딩 상태변경하도록 코드 수정

Gayun00 committed:
[Refactor] Feeds: 새 피드 로딩 시간 지연 추가

Gayun00 committed:
[Modify] Feed: 더보기 버튼 수평정렬, 피드 텍스트 마진 추가
